server/src/main/java/org/opensearch/indices/replication/RemoteStoreReplicationSource.java

```diff
@@ -93,15 +96,43 @@ public void getSegmentFiles(
         ActionListener<GetSegmentFilesResponse> listener
     ) {
         try {
-            indexShard.syncSegmentsFromRemoteSegmentStore(false, true, false);
-            listener.onResponse(new GetSegmentFilesResponse(Collections.emptyList()));
+            if (filesToFetch.isEmpty()) {
+                listener.onResponse(new GetSegmentFilesResponse(Collections.emptyList()));
+                return;
+            }
+            logger.trace("Downloading segments files from remote store {}", filesToFetch);
+            RemoteSegmentMetadata remoteSegmentMetadata = remoteDirectory.init();
+            List<StoreFileMetadata> downloadedSegments = new ArrayList<>();
+            if (remoteSegmentMetadata != null) {
+                try {
+                    indexShard.store().incRef();
+                    indexShard.remoteStore().incRef();
+                    final Directory storeDirectory = indexShard.store().directory();
+                    String segmentNFile = null;
+                    for (StoreFileMetadata fileMetadata : filesToFetch) {
+                        String file = fileMetadata.name();
+                        storeDirectory.copyFrom(remoteDirectory, file, file, IOContext.DEFAULT);
+                        downloadedSegments.add(fileMetadata);
+                        if (file.startsWith(IndexFileNames.SEGMENTS)) {
+                            assert segmentNFile == null : "There should be only one SegmentInfosSnapshot file";
+                            segmentNFile = file;
+                        }
+                    }
+                    storeDirectory.sync(downloadedSegments.stream().map(metadata -> metadata.name()).collect(Collectors.toList()));
+                } finally {
```

Comment:
Can we catch and log the exception? Also, just thinking about the exception handling: if there are 10 files to be downloaded and download of second file fails, do we get out of the download flow or continue and eventually report error?

Reply:
How is this handled today with the store? I would expect backoff & retries of the individual file.  If it still fails we need to treat it as corruption case and bail.

Reply:
We are not retrying failures on [file downloads](https://github.com/opensearch-project/OpenSearch/blob/3ee42920fa886bdf566f1bb7b1666ffa72d36f54/server/src/main/java/org/opensearch/index/shard/IndexShard.java#L4766) action. This retry would be needed in both for replication events (inside RemoteStoreReplicationSource) and for recovery (inside IndexShard.syncSegmentsFromRemoteSegmentStore). This retry logic would need more discussion. Can we take this up in separate issue ?

Reply:
Created issue https://github.com/opensearch-project/OpenSearch/issues/8885 to track retry logic.